### ISSUE TITLE
Can someone help me identify my youtube streaming buffering issue

### ISSUE THREAD

**Issue by bheeshmpita** [mpv-player/mpv#11402]
I stream videos through mpv (generally at 480p at 2x speed). I face buffering issue even at 480p. From what i have found out going through issues at github where downloading videos progresses isp speed while buffers while streaming is that ffmpeg has not implemented workarounds for youtube throttling videos >10MB size. Leaving yt-dlp current issue where all users are facing throttled speeds, i face the previously described problem. Kindly go through log file and help me identify and fix the issue. Here are some system details:

mpv 0.35.1
yt-dlp version 2023.02.17
ffmpeg version 4.4.2-0ubuntu0.22.04.1
OS Linux mint 21 xfce
CPU i3-3110m with igpu

[mpvlog.txt](https://github.com/mpv-player/mpv/files/10885398/mpvlog.txt)

**Comment by thebombzen** [comment]
yt-dlp recently implemented a fix for the >10MB issue. Update your yt-dlp.

**Comment by bheeshmpita** [comment]
@thebombzen let me rephrase my issue. I am not downloading videos with yt-dlp instead I am streaming them using mpv. When I download videos they download at full ISP speed but buffers when i stream them (even @ 480/720p). That's the issue I am looking to fix. Kindly go through log text and help me identify the issue and fix it.

**Comment by CounterPillow** [comment]
There's no way for you to fix this short of writing code for FFmpeg to make its HTTP implementation do the workarounds to evade youtube's throttling

**Comment by bheeshmpita** [comment]
@CounterPillow Thanks. Can you help me explain a little bit more what actually is happening and ffmpeg needs to do (in layman terms).

**Comment by eljamm** [comment]
This may be related to https://github.com/mpv-player/mpv/issues/10745

Playing **video-only** formats seems to throttle the speed. Here are some tests that i did:

- 398 (video-only) -> throttled
- 250 (audio-only) -> unthrottled
- 18 (video+audio) -> unthrottled
- 22 (video+audio) -> unthrottled

Merging **videos+audio** with audio doesn't seem to throttle, but with **video-only** it does:

- 22+250 -> unthrottled
- bv[height<=?480]+ba -> throttled

Versions:
- yt-dlp: 2023.03.04
- mpv: 0.35.1 + patches from https://github.com/mpv-player/mpv/pull/11398

---

Logs:
[398-throttled.txt](https://github.com/mpv-player/mpv/files/10898069/398-throttled.txt)
[250-unthrottled.txt](https://github.com/mpv-player/mpv/files/10898068/250-unthrottled.txt)
[18-unthrottled.txt](https://github.com/mpv-player/mpv/files/10898065/18-unthrottled.txt)
[22-unthrottled.txt](https://github.com/mpv-player/mpv/files/10898067/22-unthrottled.txt)
[22-250-unthrottled.txt](https://github.com/mpv-player/mpv/files/10898066/22-250-unthrottled.txt)
[bv-480-ba-throttled.txt](https://github.com/mpv-player/mpv/files/10898071/bv-480-ba-throttled.txt)

**Comment by bheeshmpita** [comment]
Thanks so much @eljamm for pointing out this to me as this basically solves my issue else I had the presumption that lower the video quality the less will it be prone to throttling. I have some other queries too, please address them if you have time:

1.  I was able to stream video without throttling by giving command from terminal but as i generally dont use terminal to watch videos can you help me by pointing out what I am missing: I tried adding `ytdl-format=22` in mpv.conf but no success, tried same in yt-dlp.conf still nothing. so basically it works from terminal but didnt otherwise. 

I had a similar issue earlier where i want to stream videos in h264 but adding stuff to mpv.conf wasnt working but when i added  `-S "codec:h264"` to yt-dlp.conf and then it worked but this time when i add `-f 22` to ut-dlp.conf it isnt loading video unthrottled.

2. Can you rephrase the #10745 (in easier terms), what I am able to understand what streaming videos other than 360 and 720p will get throttled as the audio and video file have to be merged instead streaming them collectively. So does that mean if I want to stream in 480p it will throttle?
3. Some easier understanding of   #11398 as i too get these errors and video plays without the audio.
Thanks again.

**Comment by eljamm** [comment]
2. This is the best comment that I found that describes the issue:
https://github.com/mpv-player/mpv/issues/8655#issuecomment-1199733728

Only the formats with the **https** protocol are not throttled, but they can only go up to 720p.
You can find them by using `yt-dlp -F <URL>`

Also, you can put them in order of preference. If one isn't available it will go to the next:

    mpv --ytdl-format="22/18/17/bv+ba" <URL>

---

3. This explanation is not 100% accurate, so please correct me if I'm wrong:

## yt-dlp
There are two protocols in yt-dlp for downloading files:
- __https__: Only one URL is needed.
- **dash**: Multiple URLs, also called *fragments*, are needed.

Youtube throttled downloading dash files using the https protocol, which yt-dlp fixed by changing the query[^1], but which also caused issues with how mpv handles fragments.

[^1]: https://github.com/yt-dlp/yt-dlp/issues/6400#issuecomment-1455101468 | https://github.com/yt-dlp/yt-dlp/commit/5038f6d713303e0967d002216e7a88652401c22a

## mpv
There are two types of *dash* fragments:
- **init**: Contains information about the file
- **data**: Contains actual media data and the start time to play it (duration)

Before https://github.com/mpv-player/mpv/pull/11398, the first fragment without a duration was considered as init and the ones after it as data:

> **init + data + data** + ...

However, yt-dlp splits fragments as 10MB chunks, so if the filesize is less than 10MB, it would be sent as 1 fragment. As a result, this fragment was considered to be **init**, and the data was left empty.

( As I understand, durations are used to order fragments, but if there is 1 fragment then this parameter is not included, as it's already in order. )

The fix here is to look for init fragments only if there are more than 1 fragments in total.

> **fragment + fragment + fragment** + ... --> **init + data + data** + ...

> **fragment** --> **data**

## Without fix

- **Stream size >10MB** (Usually Video): Played fine, aside from the EDL warnings.
- **Stream size <10MB** (Usually Audio): Does not play

This explains the videos with no audio.

## Applying fix

To apply this yourself, you must add the patched ytdl_hook file to `~/.config/mpv/scripts/`[^2].

    $ curl https://raw.githubusercontent.com/mpv-player/mpv/362256edbc4f95c63e69c1fa8c8dce9cc6c44288/player/lua/ytdl_hook.lua -o ~/.config/mpv/scripts/ytdl_hook.lua

You might also need to use `--ytdl=no` as mentioned in the PR.

[^2]: https://mpv.io/manual/master/#lua-scripting
